crates/ra_hir_ty/src/method_resolution.rs

```diff
@@ -381,7 +412,31 @@ fn is_valid_candidate(
                 if !data.has_self_param {
                     return false;
                 }
-                // TODO compare receiver ty
+                let substs = match m.lookup(db).container {
+                    hir_def::ContainerId::TraitId(_) => Substs::build_for_def(db, item)
+                        .push(self_ty.value.clone())
+                        .fill_with_unknown()
+                        .build(),
+                    hir_def::ContainerId::ImplId(impl_id) => {
+                        let vars =
+                            Substs::build_for_def(db, impl_id).fill_with_bound_vars(0).build();
+                        let self_ty_with_vars = db.impl_self_ty(impl_id).subst(&vars);
+                        let self_ty_with_vars =
+                            Canonical { num_vars: vars.len(), value: &self_ty_with_vars };
+                        if let Some(substs) = super::infer::unify(self_ty_with_vars, &self_ty.value)
```

Comment:
match might read better here

Reply:
that's refactored away :slightly_smiling_face: